src/parse/lexer.rs

```diff
@@ -0,0 +1,725 @@
+use crate::parse::error::{LexicalError, LexicalErrorType};
+use crate::parse::token::Tok;
+use std::char;
+
+// ideas:
+// * look up 2 / 3 char tokens like keywords
+//   * perf test
+// * track nesting per type and lex error on invalid nesting
+//   * useful?
+
+pub struct Lexer<T: Iterator<Item = (usize, char)>> {
```

Comment:
Does this work with unicode? In the hacky comments stripper we iterated across unicode grapheme clusters

Reply:
Yeah, the lexer does not need to recognize any grapheme longer than a single utf-8 byte and so if it encounters the start of a character that is longer than a byte and is:
In comment mode, it will keep consuming bytes til newline
In string mode, it will keep consuming bytes til closing quote
Not in string or comment mode, it will see it as an unrecognized token.

If it is seen as an unrecognized token Codespan is smart enough to see that the byte pointed to is a multi-byte char, and will correctly highlight it. Like this (looks better in the terminal):
```
error: Syntax error
  ┌─ /Users/a/parser_test/src/one.gleam:6:22
  │
6 │   let a = "🤠🐄 howdy" 🤠🐄
  │                        ^^ I can't figure out what to do with this character.

Hint: Is it a typo?
```

The comment blanker needs no know how many bytes a grapheme takes up because it pushes spaces (" ") to a buffer for each "character" it consumes. When it consumes a multibyte grapheme it would end up pushing the wrong number of spaces to the buffer and file locations would be off.

I actually pushed that change a few months ago. It probably could have been solved better back then, I just noticed that emoji in comments caused error messages to be off and tried switching to graphemes and it worked. :)

Reply:
Nice, thanks